### PR TITLE
udunits2 2.2.28: Build with cmake on win64

### PR DESCRIPTION
**Pull request by skupr-anaconda**
**Destination channel:** defaults

### Links

- [PKG-4302](https://anaconda.atlassian.net/browse/PKG-4302) 
- [Upstream repository](https://github.com/Unidata/UDUNITS-2)
- Source files https://downloads.unidata.ucar.edu/udunits/2.2.28/udunits-2.2.28.tar.gz

### Explanation of changes:

- Remove obsolete patches
- Add a new [recipe/fix_output_dll_name.patch](https://github.com/AnacondaRecipes/udunits2-feedstock/pull/2/files#diff-216733931825e5d11d1ebc6ae51c009f259f54eb2a02581b7ba098ab4fee8ad0) to fix output names for .dll, .lib, .exp.
- Add activation scripts
- Rename the package on `win`: the old name `m2w64-udunits2` is [used by R packages](https://metayaml-conda.fly.dev/metayaml/reverse_dependencies?name=m2w64-udunits2) with `r-base` **<4.3.1** but a new name `udunits2` will be used by R packages with `r-base` **4.3.1+** on `win`.
- Update source/urls
- Reset the build number to `0`
- Skip `s390x` as we didn't build it previously and it seems there are compatibility issues
- Modify `detect_binary_files_with_prefix: true` as conda-forge did
- Update build dependencies, use the Windows compiler on `win64`, add `texinfo`/`m2-texinfo` for `makeinfo` functionality
- Use `expat` from **defaults**
- Remove `{{n}}gcc-libs` on `win` as we do not rely on `m2w64 toolchain`
- Add the command `if not exist "%LIBRARY_PREFIX%\\share\\udunits\\udunits2.xml" exit 1  # [win]`
- Add more test commands for `win`
- Remove `conda inspect` commands
- Fix `home` url and `license` name, add dev & doc urls, `description`, `license_family`

### Notes:

- The previous attempt with `m2w64-toolchain` and `make` was https://github.com/AnacondaRecipes/udunits2-feedstock/pull/1
- I've tested the downstream `r-udunits2` and it works fine.

[PKG-4302]: https://anaconda.atlassian.net/browse/PKG-4302?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/bld.bat [hunk 1]
```diff
@@ -1,9 +1,43 @@
...
 if errorlevel 1 exit 1
-exit 0
+
+nmake libudunits2 || exit 1
+nmake udunits2 || exit 1
+
+mkdir %SCRIPTS%
+
+copy prog\udunits2.exe %SCRIPTS%\udunits2.exe || exit 1
+copy lib\udunits2.dll %SCRIPTS%\udunits2.dll || exit 1
```
Comment (JeanChristopheMorinPerso) [inline recipe/bld.bat]: Should this be
```suggestion
copy prog\udunits2.exe %LIBRARY_BIN%\udunits2.exe || exit 1
copy lib\udunits2.dll %LIBRARY_BIN%\udunits2.dll || exit 1
```
?

Comment (skupr-anaconda) [reply recipe/bld.bat]: Yeah, I've made these changes locally today. I'll push them

Comment (skupr-anaconda) [reply recipe/bld.bat]: Fixed

**Comment by skupr-anaconda** [comment]
@JeanChristopheMorinPerso Is it possible to [to set/patch the executable name in CMake?](https://stackoverflow.com/questions/24395517/how-to-set-the-executable-name-in-cmake). Because the DLL name `udunits2.dll` clashes with where **R's** `udunits2.dll` in `C:/Users/dev-admin/conda-bld/r-udunits2_1711363840246/_h_env/Lib/R/library/00LOCK-work/00new/udunits2/libs/x64`.  `r-udunits2` is assuming an MSYS2 library and so also builds a `udunits2.dll` shim library dependent on the `udunits2.dll` 

But `libudunits2-1.dll` seems will work fine (not 100% sure).

**Comment by JeanChristopheMorinPerso** [comment]
>Is it possible to [to set/patch the executable name in CMake?](https://stackoverflow.com/questions/24395517/how-to-set-the-executable-name-in-cmake)

In CMake, we can set the [OUTPUT_NAME](https://cmake.org/cmake/help/latest/prop_tgt/OUTPUT_NAME.html) using `set_target_properties` like this:
```
set_target_properties(my_target PROPERTIES OUTPUT_NAME "better_name")
```

**Comment by skupr-anaconda** [comment]
> In CMake, we can set the [OUTPUT_NAME](https://cmake.org/cmake/help/latest/prop_tgt/OUTPUT_NAME.html) using `set_target_properties` like this:
> 
> ```
> set_target_properties(my_target PROPERTIES OUTPUT_NAME "better_name")
> ```

If I patch [CMakeLists.txt](https://github.com/Unidata/UDUNITS-2/blob/master/CMakeLists.txt#L49-L54) with:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index fbc2fde..b497fb5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -49,6 +49,7 @@ IF(MSVC)
     MESSAGE(STATUS "Building Windows DLL: ${ENABLE_DLL}")
     IF(ENABLE_DLL)
        SET(BUILD_DLL ON CACHE BOOL "")
+       set_target_properties("udunits2" PROPERTIES OUTPUT_NAME "udunits2-1")
        ADD_DEFINITIONS(-DDLL_UDUNITS2)
        ADD_DEFINITIONS(-DDLL_EXPORT)
     ENDIF()
```
then the error is:

```
CMake Error at CMakeLists.txt:52 (set_target_properties):
  set_target_properties Can not find target to add properties to: udunits2
```

@M-Waszkiewicz-Anaconda @JeanChristopheMorinPerso Am I doing something wrong?

**Comment by skupr-anaconda** [comment]
@JeanChristopheMorinPerso `set(CMAKE_RELEASE_POSTFIX "-1")`  will fix the issue

**Comment by JeanChristopheMorinPerso** [comment]
@skupr-anaconda You have to do it after the target gets defined. https://github.com/Unidata/UDUNITS-2/blob/master/lib/CMakeLists.txt#L62-L66

**Comment by JeanChristopheMorinPerso** [comment]
>`set(CMAKE_RELEASE_POSTFIX "-1")` will fix the issue

Oh, I didn't know this trick. Thanks for sharing.

I have another thing in mind and I'd like to see what you think about it. I think it would be preferable if we keep udunits2's library name as is and instead modify R's `udunits2.dll` to have a different name?

When I look at https://github.com/pacificclimate/Rudunits2/blob/master/configure.ac#L60 for example, I see it wants to link `udunits2`. If you modify `udunits2` to be `udunits2-1`, then you'll have to edit Rudunits2 to match the new library name, only on Windows.

Do you think it would be feasible?

**Comment by skupr-anaconda** [comment]
> @skupr-anaconda You have to do it after the target gets defined. https://github.com/Unidata/UDUNITS-2/blob/master/lib/CMakeLists.txt#L62-L66

I've renamed outputs because adding new names won't work:
```
set_target_properties(libudunits2 PROPERTIES LIBRARY_OUTPUT_NAME udunits2)
set_target_properties(libudunits2 PROPERTIES ARCHIVE_OUTPUT_NAME udunits2)
set_target_properties(libudunits2 PROPERTIES RUNTIME_OUTPUT_NAME udunits2)
set_target_properties(libudunits2 PROPERTIES LIBRARY_OUTPUT_NAME udunits2-1)
set_target_properties(libudunits2 PROPERTIES ARCHIVE_OUTPUT_NAME udunits2-1)
set_target_properties(libudunits2 PROPERTIES RUNTIME_OUTPUT_NAME udunits2-1)
```
and the result will be that  `lib\udunits2.dll` isn't available:
```
[ 93%] Building C object lib/CMakeFiles/libudunits2.dir/udunits_snprintf.c.obj
udunits_snprintf.c
[100%] Linking C shared library udunits2-1.dll
[100%] Built target libudunits2

(%BUILD_PREFIX%) %SRC_DIR%\build>nmake udunits2   || exit 1

Microsoft (R) Program Maintenance Utility Version 14.16.27051.0
Copyright (C) Microsoft Corporation.  All rights reserved.

[ 84%] Built target libudunits2
[ 89%] Building C object prog/CMakeFiles/udunits2.dir/udunits2.c.obj
udunits2.c
[ 94%] Building C object prog/CMakeFiles/udunits2.dir/XGetopt.c.obj
XGetopt.c
[100%] Linking C executable udunits2.exe
[100%] Built target udunits2

(%BUILD_PREFIX%) %SRC_DIR%\build>mkdir %PREFIX%\Scripts

(%BUILD_PREFIX%) %SRC_DIR%\build>copy prog\udunits2.exe %PREFIX%\Library\bin\udunits2.exe   || exit 1
        1 file(s) copied.

(%BUILD_PREFIX%) %SRC_DIR%\build>copy lib\udunits2.dll %PREFIX%\Library\bin\udunits2.dll   || exit 1
The system cannot find the file specified.
```

**Comment by skupr-anaconda** [comment]
I've tested `r-udunits2` and it works fine if I patch its `bld.bat` like this:
```
set PKG_LIBS=-L$(R_HOME)/../../Library/bin -ludunits2-2 -lexpat
sed -i.bak "s/^PKG_LIBS/#PKG_LIBS/" src/Makevars.ucrt

"%R%" CMD INSTALL --build . --configure-args="%PKG_LIBS%"
```

**Comment by JeanChristopheMorinPerso** [comment]
> I've tested `r-udunits2` and it works fine if I patch its `bld.bat` like this:
> 
> ```
> set PKG_LIBS=-L$(R_HOME)/../../Library/bin -ludunits2-2 -lexpat
> sed -i.bak "s/^PKG_LIBS/#PKG_LIBS/" src/Makevars.ucrt
> 
> "%R%" CMD INSTALL --build . --configure-args="%PKG_LIBS%"
> ```

But what if there is ever a need to use udunits2 in other packages? We'd probably have to patch or tweak them too. SHouldn't we fix the problematic package (either r-udunits2 or r-base or whatever other package that also has a `udunits2.dll`) instead of making defaults's udunits2 an exception and have to deal with that exception in potentially multiple places?

**Comment by skupr-anaconda** [comment]
Only four packages `r-udunits2`, `r-units`, `r-tikzdevice`, and `r-rnetcdf` require `udunits2`. We should patch them anyway. Seems there is no other way to unblock them

**Comment by JeanChristopheMorinPerso** [comment]
Where does the conflicting `udunits2.dll` come from?

**Comment by skupr-anaconda** [comment]
> Where does the conflicting `udunits2.dll` come from?

`r-base` creates it in a temporary folder `C:/Users/dev-admin/conda-bld/r-udunits2_1711112398177/_h_env/Lib/R/library/00LOCK-work/00new/udunits2/libs/x64/udunits2.dll` and it fails https://github.com/AnacondaRecipes/udunits2-feedstock/pull/1#discussion_r1535545844

**Comment by JeanChristopheMorinPerso** [comment]
> > Where does the conflicting `udunits2.dll` come from?
> 
> `r-base` creates it in a temporary folder `C:/Users/dev-admin/conda-bld/r-udunits2_1711112398177/_h_env/Lib/R/library/00LOCK-work/00new/udunits2/libs/x64/udunits2.dll` and it fails [#1 (comment)](https://github.com/AnacondaRecipes/udunits2-feedstock/pull/1#discussion_r1535545844)

Shouldn't that be removed instead of patching udunits2?

**Comment by skupr-anaconda** [comment]
> > > Where does the conflicting `udunits2.dll` come from?
> > 
> > 
> > `r-base` creates it in a temporary folder `C:/Users/dev-admin/conda-bld/r-udunits2_1711112398177/_h_env/Lib/R/library/00LOCK-work/00new/udunits2/libs/x64/udunits2.dll` and it fails [#1 (comment)](https://github.com/AnacondaRecipes/udunits2-feedstock/pull/1#discussion_r1535545844)
> 
> Shouldn't that be removed instead of patching udunits2?

It's an internal r-base's process we can't change it. @ifitchet investigated it in detail and knows it better

**Comment by JeanChristopheMorinPerso** [comment]
From what I understand, `00LOCK-work` is a leftover from a failed build. @ifitchet do you have more info on why this happens to be in the `r-base` package?

**Comment by ifitchet** [comment]
@JeanChristopheMorinPerso `00LOCK-work/00new` is a temporary directory used by `r-base` when constructing R libraries from source.  As you say, you would normally only see it when a build fails.  It is being used during a build when you see:

```
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
```

These shim libraries are used for every compilable R library and they exist (eventually) in `${PREFIX}/lib/R/library/FOO/lib/x64/FOO.dll` (for some `FOO`), and, I would presume contain all the code for providing the R interfaces and for converting R data structures to/from calls to the underlying implementation library (in C or C++ ...).

The real question is, how does R chose its shim library names?  I tried to follow the R code through but didn't get very far.  You might assume, not unreasonably, that given an R library name of `FOO` the R interpreter will be looking for `FOO.dll` in the above location.

We can't rename the R library `FOOx` because R users will get upset.  Can we keep the R library name of `FOO` and somehow get it to name the DLL `FOOx.dll`?  I haven't figured out how.

Ultimately, though, if the R library is `FOO.dll` and the underlying implementation library is `foo.dll` then the Windows DLL search algorithm will see that R's `FOO.dll` is dependent on `foo.dll` and then immediately find `FOO.dll` in the same directory (thus failing to find the actual implementation function(s)).

Why does MSYS2 work?  MSYS2 by and large uses `configure` and `configure` seems to generate libraries called `libfoo-x.dll` which the R library, `FOO.dll` is able to link against _and subsequently find_ because the physical file name is different.

(I don't know if that answers your question. 😆 )

**Comment by JeanChristopheMorinPerso** [comment]
Thanks @ifitchet. What I'm curious to know is why `00LOCK-work/00new` exists. For example, when I look at https://conda-metadata-app.streamlit.app/?q=pkgs%2Fr%2Flinux-64%2Fr-base-4.3.1-h1ae530e_0.conda, I don't see any temp directory. The same is true on conda-forge, see https://conda-metadata-app.streamlit.app/?q=conda-forge%2Fwin-64%2Fr-base-4.1.3-hdca333a_12.conda.

So to me the real question is why would our new msys2 based r-base package have `00LOCK-work/00new` directories while packages on conda-forge and the r channel don't?

**Comment by ifitchet** [comment]
@JeanChristopheMorinPerso the temporary directories are part of the R library build process -- quite _why_ they need this step, I don't know -- and *all* platforms will use them.  There's nothing special about Windows, here.

But they are only transient and, assuming everything worked, the shim library is moved to its permanent location `lib/R/library/FOO/lib/x64/FOO.dll`, the temporary directory is removed and the build process continues with the next step (whatever that is, I forget).

So, if you watch a `linux-64` build carefully, you'll see the same `00LOCK-work/00new` directories come and go.  (I assume the `00` can be `01`, `02` etc. depending on circumstances.)

Come the final packaging, then, there's nothing regarding `00LOCK-work/00new` because it's been deleted.

**Comment by JeanChristopheMorinPerso** [comment]
>Come the final packaging, then, there's nothing regarding 00LOCK-work/00new because it's been deleted.

You just lost me there. If it's deleted, why is it a problem when building r-udunits2?

**Comment by ifitchet** [comment]
@JeanChristopheMorinPerso because, on Windows, the R library `r-udunits2` creates a shim library called `udunits2.dll` which is looking for the implementation library (from defaults) called `udunits2.dll` (the same name).  

_Windows_'s DLL search algorithm says that if an executable is trying to load a DLL then start looking for that DLL in the same directory as the executable: so the R shim library `udunits2.dll` (in `.../00LOCK-work/00new/...`) will start looking for the implementation library (which is using the same name) in its own directory and finds itself (not the implementation library which is elsewhere).

So that step of the R build process fails (loading from a temporary directory) and we so happen to see `00LOCK-work/00new` exposed.

For other R libraries either the R library is called something "safe", eg. `r-rFOO` implying a DLL name of `rFOO.dll` which is safe to find the implementation library `foo.dll` or the implementation library is called something different anyway (like `foo-x.dll` as conjured by `configure`).

**Comment by JeanChristopheMorinPerso** [comment]
>We can't rename the R library FOOx because R users will get upset. Can we keep the R library name of FOO and somehow get it to name the DLL FOOx.dll? I haven't figured out how.

If I understand correctly, https://github.com/pacificclimate/Rudunits2/blob/master/NAMESPACE#L2 could be patched to load a different shim library (for example, `udunits2-r` instead of `udunits2`). See https://cran.r-project.org/doc/manuals/R-exts.html#useDynLib-1.

From what I see, `lib/R/library/udunits2/R/udunits2` contains
```R
local({
    info <- loadingNamespaceInfo()
    pkg <- info$pkgname
    ns <- .getNamespace(as.name(pkg))
    if (is.null(ns))
        stop("cannot find namespace environment for ", pkg, domain = NA);
    dbbase <- file.path(info$libname, pkg, "R", pkg)
    lazyLoad(dbbase, ns, filter = function(n) n != ".__NAMESPACE__.")
})
```
so it basically loads the NAMESPACE file `lib/R/library/udunits2/NAMESPACE` (or something like that), which looks like
```
exportPattern("^ud.*")
useDynLib(udunits2, R_ut_are_convertible, R_ut_convert, R_ut_get_name, R_ut_get_symbol, R_ut_is_parseable, R_ut_set_encoding, R_ut_init)
```

`useDynLib(udunits2, R_ut_are_convertible, R_ut_convert, R_ut_get_name, R_ut_get_symbol, R_ut_is_parseable, R_ut_set_encoding, R_ut_init)` I think means "Use the functions `R_ut_are_convertible`, `R_ut_convert`, etc from the library named `udunits2`".

**Comment by JeanChristopheMorinPerso** [comment]
@skupr-anaconda @ifitchet any update on this?

**Comment by ifitchet** [comment]
> any update on this?

No.  And I guess we have little incentive to progress it at this point.

**Comment by JeanChristopheMorinPerso** [comment]
@ifitchet should we close it then?

**Comment by skupr-anaconda** [comment]
Seems like we can close it now

**Pull request closed without merging**